packages/solid-query/src/createBaseQuery.ts

```diff
@@ -35,46 +42,99 @@ export function createBaseQuery<
   const queryClient = createMemo(() =>
     useQueryClient({ context: options().context }),
   )
-  const emptyData = Symbol('empty')
 
   const defaultedOptions = queryClient().defaultQueryOptions(options())
   defaultedOptions._optimisticResults = 'optimistic'
   const observer = new Observer(queryClient(), defaultedOptions)
 
+  const emptyData = Symbol('empty')
+
+  let resolver: (value: TData | undefined) => void
+  let queryResolver: (value: DehydratedState) => void
+
+  // This resource will be used in a Server environment to
+  // dehydrate the queryClient when the query is
+  // pre fetched on the server. It will always be undefined
+  // if an observer is mounted on client.
+  const [queryResource] = createResource<DehydratedState | undefined>(() => {
+    return new Promise((resolve) => {
+      if (!isServer) {
+        resolve(undefined)
+      }
+      queryResolver = resolve
+    })
+  })
+
+  // If queryResource is defined,
+  // This means that the query was fetched on the server!
+  // We need to hydrate the queryClient with the query from
+  // the server before we set up the observer and its results.
+  if (queryResource()) {
+    hydrate(queryClient(), queryResource())
+  }
+
   const [state, setState] = createStore<QueryObserverResult<TData, TError>>(
     observer.getOptimisticResult(defaultedOptions),
   )
 
-  const [dataResource, { refetch, mutate }] = createResource<TData | undefined>(
+  const [dataResource, { mutate, refetch }] = createResource<TData | undefined>(
     () => {
       return new Promise((resolve) => {
-        if (!(state.isFetching && state.isLoading)) {
-          if ((unwrap(state.data) as TData | typeof emptyData) === emptyData) {
-            resolve(undefined)
+        if (isServer) {
+          if (!(state.isFetching && state.isLoading)) {
+            const dehydratedClient = dehydrate(queryClient())
+            queryResolver(dehydratedClient)
+            resolve(unwrap(state.data))
+          }
+          // We only resolve the data resource
+          // when the query observer finds a result
+          // This function will be called inside the observer
+          // subscription function.
+          resolver = resolve
+        } else {
+          if (!(state.isFetching && state.isLoading)) {
+            if (
+              (unwrap(state.data) as TData | typeof emptyData) === emptyData
+            ) {
+              resolve(undefined)
+            }
+            resolve(unwrap(state.data))
           }
-          resolve(unwrap(state.data))
         }
       })
```

Comment:
Any way that we could expose the createResource `deferStream` option? Seems to be working very nicely (and could make for a neat extra route in the solid-start example, to show off deferred and non-deferred streaming), but I'm not sure what pattern ya'll are using to expose adapter specific options up on `createQuery`.

Reply:
However, I did notice this warning. Does not seem to cause any issues in this simple example, but I suspect it will cause random bugs in a more complex / real world application unless resolved. The same query done with a regular createResource (rather than through solid-query) does not result in the hydration warning.

<img width="662" alt="Screenshot 2023-01-13 at 3 40 51 PM" src="https://user-images.githubusercontent.com/847542/212415216-ff1bef63-57be-4a22-bd4d-60291b51b978.png">

Reply:
> However, I did notice this warning. Does not seem to cause any issues in this simple example, but I suspect it will cause random bugs in a more complex / real world application unless resolved. The same query done with a regular createResource (rather than through solid-query) does not result in the hydration warning.
> 
> <img alt="Screenshot 2023-01-13 at 3 40 51 PM" width="662" src="https://user-images.githubusercontent.com/847542/212415216-ff1bef63-57be-4a22-bd4d-60291b51b978.png">

Hello! Thank you so much for taking out time to review the changes. Really appreciate it ❤️  Interesting! I have once seen this issue when I didn't use a signal in the JSX. But I cannot seem to recreate this issue again. I have a Stackblitz example if you would like to fork to recreate the warning? I can then take a look to see why that happens. Maybe just an overly prescriptive warning? Will definitely confirm what's causing it. https://stackblitz.com/edit/github-bo2rcr?file=src%2Froutes%2Findex.tsx

Reply:
> Any way that we could expose the createResource `deferStream` option? Seems to be working very nicely (and could make for a neat extra route in the solid-start example, to show off deferred and non-deferred streaming), but I'm not sure what pattern ya'll are using to expose adapter specific options up on `createQuery`.

Yeah we can definitely expose `deferStream` as an option to `createQuery`/`createinfiniteQuery`. I tried futzing around with streaming SSR but could not figure out how `deferStream` works. Do you have an example how `deferStream` works in resources? That would be immensely helpful

Reply:
> Do you have an example how deferStream works in resources? That would be immensely helpful

For sure - try adjusting the example to https://github.com/TanStack/query/pull/4803#discussion_r1069858512 + updating `entry-server.tsx` to use `renderStream` rather than `renderAsync`. Then if you turn off javascript in your browser you should see the difference quite clearly when toggling "deferStream" true/false for that `post` query.

> Hello! Thank you so much for taking out time to review the changes. Really appreciate it ❤️ Interesting! I have once seen this issue when I didn't use a signal in the JSX. But I cannot seem to recreate this issue again.

Oh no problem, thanks for getting the ball rolling! This DX is going to be so great, I'm pretty excited about it.

Ah, you'll see the hydration key issue if you adjust to use `renderStream` & set `deferStream` to true (and make sure javascript back on in browser). 

It think it's something in the solid-query implementation, because if one adjusts to just use `createResource` directly to fetch the post, ala the example below, the hydration key warning goes away 🤔.

```ts
function Post({ postId }: { postId: Accessor<number> }) {
  const [postData] = createResource(async () => {
    const response = await fetch(
      'https://jsonplaceholder.typicode.com/posts/' + postId(),
    ).then((res) => res.json())

    // simulate extra latency to make things like streaming behavior more clear
    await new Promise((r) => setTimeout(r, 1000))

    return [response] as PostData[];
  }, { 
    deferStream: true,
  })

  return (
    <div>
      <For each={postData()}>
        {(post) => (
          <div>
            <h3>{post.title}</h3>
            <p>{post.body}</p>
          </div>
        )}
      </For>
    </div>
  )
}
```

Reply:
A little update on this. I think I figured out why we get the `hydration key` warning. It's because (I think) it's trying to find a dom node where `queryResource` is being used in the component tree. But it's only used internally by createQuery!

Reply:
Oooh! I think I fixed it 🥳 🥳 🥳

Reply:
Unfortunately, I'm still seeing the hydration error. BUT, I think I know why now - mind if I open a separate PR with an adjusted approach and explanation (primarily, consolidates to a single `createResource` call)? Can then compare/contrast/experiment with it, and either incorporate the changes here, roll w the other PR, or find another approach if it's not valid - I don't mind either way.

Reply:
Which version of @tanstack/query are you using? Is it the one with the Code Sandbox link above?

Reply:
Also, yes! PRs are always welcome 😄

Reply:
I think I know now. Try removing the fallback on your suspense boundaries. You wont see the hydration warning anymore 😄 That's weird, it's expecting the first pass render to have the data resource be undefined on the client when hydrating!

Reply:
I have a discussion on the solid-start discussion channel. If you take a look this hydration key warning is being experienced by a lot of folks (for people not using tanstack query). It doesn't affect anything. This maybe something solid-start needs to fix

Reply:
I'll join it :).

Popped the WIP up over here if you're interested -> #4840.